core/components/bitvector.hpp

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: 2024 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_COMPONENTS_BITVECTOR_HPP_
+#define GKO_CORE_COMPONENTS_BITVECTOR_HPP_
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/intrinsics.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+namespace gko {
+
+
+template <typename IndexType>
+class device_bitvector {
+public:
+    using index_type = IndexType;
+    using storage_type = uint32;
+    constexpr static int block_size = CHAR_BIT * sizeof(storage_type);
+
+    constexpr device_bitvector(const storage_type* bits,
+                               const index_type* ranks, index_type size)
+        : bits_{bits}, ranks_{ranks}, size_{size}
+    {}
+
+    constexpr index_type size() const { return size_; }
+
+    constexpr index_type num_blocks() const
+    {
+        return (size() + block_size - 1) / block_size;
+    }
+
+    constexpr bool get(index_type i) const
+    {
+        assert(i >= 0);
+        assert(i < size());
+        const auto block = i / block_size;
+        const auto local = i % block_size;
+        return bool((bits_[block] >> local) & 1);
+    }
+
+    constexpr index_type rank(index_type i) const
```

Comment:
we could potentially also add `rank_inclusive` that adds 1 if the bit is set, then we can use that to represent row_ptrs